File: tier4_expert/monotonic_stack/problems/p739_daily_temperatures.py
"""P739: Daily Temperatures (Medium)

TODO: Implement solve() below.
Hint: Use a monotonic decreasing stack to track indices of warmer days.
"""
import sys
sys.path.insert(0, ".")
from src.utils import Problem, TestCase

class Solution(Problem):
    name = "739. Daily Temperatures"
    test_cases = [
        TestCase(input=[73,74,75,71,69,72,76,73], expected=[1,1,4,2,1,1,0,0], label="example 1"),
        TestCase(input=[30,40,50,60], expected=[1,1,1,0], label="example 2"),
        TestCase(input=[30,60,90], expected=[1,1,0], label="example 3"),
    ]

    def solve(self, temperatures: list[int]) -> list[int]:
        raise NotImplementedError("TODO: Implement solve(self, temperatures: list[int]) -> list[int]")

if __name__ == "__main__":
    Solution().run()

File: tier4_expert/monotonic_stack/solutions/p739_daily_temperatures.mjs
/**
 * P739: Daily Temperatures (Medium)
 * https://leetcode.com/problems/daily-temperatures/
 * Topics: Array, Stack, Monotonic Stack
 *
 * Given an array of integers temperatures represents the daily temperatures, return an array answer such that answer[i] is the number of days you have to wait after the ith day to get a warmer temperature. If there is no future day for which this is possible, keep answer[i] == 0 instead.
 *
 * Example 1:
 *     Input: temperatures = [73,74,75,71,69,72,76,73]
 *     Output: [1,1,4,2,1,1,0,0]
 *
 * Example 2:
 *     Input: temperatures = [30,40,50,60]
 *     Output: [1,1,1,0]
 *
 * Example 3:
 *     Input: temperatures = [30,60,90]
 *     Output: [1,1,0]
 *
 * Constraints:
 *     - 1 <= temperatures.length <= 105
 *     - 30 <= temperatures[i] <= 100
 *
 * Hint: Use a monotonic decreasing stack to track indices of warmer days.
 */

import { readLine, readInts, readInt, writeInt, writeInts, writeString, writeBool } from '../../../wasm_libs/js/io.mjs';

function solve(temperatures) {
  const n = temperatures.length;
  const answer = new Array(n).fill(0);
  const stack = [];
  for (let i = 0; i < n; i++) {
    while (stack.length > 0 && temperatures[i] > temperatures[stack[stack.length - 1]]) {
      const j = stack.pop();
      answer[j] = i - j;
    }
    stack.push(i);
  }
  return answer;
}

const temperatures = readInts();
writeInts(solve(temperatures));
